lib/Sema/CSDiag.cpp

```diff
@@ -5421,8 +5431,96 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
     // Type check the function subexpression to resolve a type for it if
     // possible.
     fnExpr = typeCheckChildIndependently(callExpr->getFn());
-    if (!fnExpr)
-      return true;
+    if (!fnExpr) {
+      auto hadError = [&]() {
+        return CS.TC.Diags.hadAnyError();
+      };
+      auto UDE = dyn_cast<UnresolvedDotExpr>(callExpr->getFn());
+      if (!UDE)
+        return hadError();
+
+      auto baseExpr = UDE->getBase();
+      // Offer fix-it for using `subscript` instead of the operator.
+      // FIXME: This doesn't check generic parameters
+      if (baseExpr && UDE->getName().getBaseName() == "subscript") {
+        auto baseType = CS.getType(baseExpr);
+        // Look up subscript declarations and return if there are none.
+        auto lookup = CS.lookupMember(baseType->getRValueType(),
+                                      DeclName(DeclBaseName::createSubscript()));
+        if (lookup.empty())
+          return hadError();
+
+        auto argExpr = typeCheckChildIndependently(callExpr->getArg(), Type(),
+                                                   CTP_CallArgument);
+        if (!argExpr)
+          return hadError();
+
+        SmallVector<Identifier, 2> scratch;
+        ArrayRef<Identifier> argLabels = callExpr->getArgumentLabels(scratch);
+        SmallVector<OverloadChoice, 2> choices;
+        auto hasTrailClosure = callArgHasTrailingClosure(argExpr);
+
+        for (auto candidate : lookup)
+          choices.push_back(OverloadChoice(baseType, candidate.getValueDecl(),
+                                           UDE->getFunctionRefKind()));
+
+        CalleeCandidateInfo candidates(baseType, choices,
+                                       hasTrailClosure, CS, true);
+
+        auto argType = CS.getType(argExpr);
+        auto params = swift::decomposeArgType(argType, argLabels);
+
+        using Closeness = CalleeCandidateInfo::ClosenessResultTy;
+
+        candidates.filterList([&](UncurriedCandidate candidate) -> Closeness {
+          auto candFuncType = candidate.getUncurriedFunctionType();
+          if (!candFuncType)
+            return {CC_GeneralMismatch, {}};
+
+          auto candParams = candFuncType->getParams();
+
+          if (params.size() != candParams.size())
+            return {CC_GeneralMismatch, {}};
+
+          for (unsigned i = 0, e = params.size(); i < e; i ++) {
+            auto type1 = params[i].getType()->
+              getUnlabeledType(CS.getASTContext());
+            auto type2 = candParams[i].getType()->
+              getUnlabeledType(CS.getASTContext());
```

Comment:
Otherwise `(T, U)` and `(arg: T, U)`, for instance, aren't equal. If there is a better way to check for equality ignoring labels, I'll be glad to hear.

Reply:
Sorry, I missed this thread, tuples`(T, U)` and `(arg: T, U)` might be considered equal because we have an implicit conversion for that (if parameter is unlabeled),  but e.g. `(a: Int, b: Int)` vs. `(a: Int, c: Int)` should not be, so I think it's incorrect to always trip the labels from parameters.

Reply:
So... Check tuple arguments for compatibility. Is there an implicit conversion check for that? Or
```
if (unlabeledType1 == Type1 && unlabeledType1 == unlabeledType2) 
```
This might become tricky with generics since they are compound types.

Reply:
I think we should just check without unlabeling, that would be the safest option, because there is no easy way to properly check for implicit conversions.